feat: Enable Apps for Teams & Orgs [CAL-1782]

joeauyeung opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

This PR does the following:
- Allows users to install apps for either themselves or teams if they have the permission
- Show personal and team apps on the installed page
- Enable apps on the event type apps page for teams
- Allow apps for managed event types

Followup:
- Video Apps support for teams.

https://www.loom.com/share/ab4033942aac43a084cd077551b1b3c9

Fixes # (issue)

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

## Type of change

<!-- Please delete bullets that are not relevant. -->

- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

- [x] Create a team
- [x] On the app store page, install an app
    - [x] A drop down should appear to install for either the user or a team
    - [x] Install the app for both instances
- [x] Go to a specific app page
    - [x] When install the app the dropdown should appear for the user or a team
    - [x] This time install the app for just a team
- [x] On the installed app pages
    - [x] Installed apps for the user and teams should appear
- [x] Go to a team event type and navigate to the apps page
    - [x] It should show credentials for the user and the team
    - [x] When enabling one it should save and not enable both instances of the app

## Mandatory Tasks
- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

<!-- Please remove all the irrelevant bullets to your PR -->
- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my PR needs changes to the documentation
- I haven't checked if my changes generate no new warnings
- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes


## I recommend adding feedback here as it is much easier to keep a track 
###  Feedback by Hariom
#### Blocking
- [x] Broken UI for Remove App Dialog. It is broken in production as well. [So, created a separate issue.](https://github.com/calcom/cal.com/issues/9846)
- [x] Can't install an App for a team(after i have installed it for user)
- [x] Revert [this](https://github.com/calcom/cal.com/pull/9337/files#r1248819904), [this](https://github.com/calcom/cal.com/pull/9337/files#r1248862915) and [this](https://github.com/calcom/cal.com/pull/9337/files#r1250164219) change 
- [x] [Possible 500](https://github.com/calcom/cal.com/pull/9337/files#r1248873555)
- [x] Clarity on https://github.com/calcom/cal.com/pull/9337/files#r1250136887
- [x] [Permission related issues](https://github.com/calcom/cal.com/pull/9337#issuecomment-1617307264)
- TODO
	- [x] Zoom installation because state is being passed now
	- [x] Stripe booking payment test for Team App installation and user App installation. Refer to this [change](https://github.com/calcom/cal.com/pull/9337/files#r1248874162)
	- [x] Test Managed Event Types.

leog committed:
Initial commit

leog committed:
Adding feature flag

leog committed:
Merge branch 'main' into feat/organizations

leog committed:
Merge branch 'main' into feat/organizations

leog committed:
Merge branch 'main' into feat/organizations

leog committed:
Merge branch 'main' into feat/organizations

joeauyeung committed:
Add schema relation for teams and credentials

joeauyeung committed:
feat: Orgs Schema Changing `scopedMembers` to `orgUsers` (#9209)

* Change scopedMembers to orgMembers

* Change to orgUsers

joeauyeung committed:
Create getUserAdminTeams function & tRPC endpoint

joeauyeung committed:
Get user admin teams on app store page

joeauyeung committed:
Create UserAdminTeams type

leog committed:
Merge branch 'main' into feat/organizations

joeauyeung committed:
Add user query to getUserAdminTeams

leog committed:
Letting duplicate slugs for teams to support orgs

leog committed:
Merge branch 'main' into feat/organizations

leog committed:
Covering null on unique clauses

joeauyeung committed:
Add dropdown to install button on app store

leog committed:
Supporting having the orgId in the session cookie

leog committed:
Merge branch 'main' into feat/organizations

joeauyeung committed:
Merge branch 'feat/organizations' into feat/apps-for-orgs

joeauyeung committed:
On app page, only show dropdown if there are teams

joeauyeung committed:
Add teamId to OAuth state

joeauyeung committed:
Create team credential for OAuth flow

joeauyeung committed:
Create team credential for GCal

joeauyeung committed:
Add create user or team credential for Stripe

joeauyeung committed:
Create webex credentials for users or teams

joeauyeung committed:
Fix type error on useAddAppMutation

joeauyeung committed:
Hubspot create credential on user or team

joeauyeung committed:
Zoho create create credential for user or team

joeauyeung committed:
Zoom create credentials on user or team

joeauyeung committed:
Salesforce create credential on user or teams

joeauyeung committed:
OAuth create credentials for user or teams

leog committed:
Merge branch 'main' into feat/organizations

joeauyeung committed:
Merge branch 'feat/organizations' into feat/apps-for-orgs

joeauyeung committed:
Revert Outlook changes

joeauyeung committed:
Revert GCal changes

joeauyeung committed:
Default app instal, create credential on user or team

joeauyeung committed:
Add teamId to credential creation

joeauyeung committed:
Disable installing for teams for calendars

leog committed:
Merge branch 'main' into feat/organizations

joeauyeung committed:
Include teams when querying installed apps

joeauyeung committed:
Render team credentials on installed page

joeauyeung committed:
Uninstall team apps

joeauyeung committed:
Type fix on app card

joeauyeung committed:
Add input to include user in teams query

joeauyeung committed:
Add dropdown to install app page  for user or team

leog committed:
Merge branch 'main' into feat/organizations

joeauyeung committed:
Type fixes on category page

joeauyeung committed:
Install app from eventType page to user or team

joeauyeung committed:
Render user and team apps on event type app page

joeauyeung committed:
Render user and team apps on event type app page

leog committed:
Merge branch 'main' into feat/organizations

joeauyeung committed:
Add credentialOwner to eventTypeAppCard types

joeauyeung committed:
Type fixes

joeauyeung committed:
Create hook to check if app is enabled

Udit-takkar committed:
feat: organization event type filter (#9253)

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

joeauyeung committed:
Clean up console.logs

joeauyeung committed:
Fix useIsAppEnabled by returning not an array

leog committed:
Merge branch 'main' into feat/organizations

leog committed:
Missing changes to support orgs schema changes

joeauyeung committed:
Convert event type apps to useIsAppEnabled

joeauyeung committed:
Abstract credential owner type

joeauyeung committed:
Remove console.logs

joeauyeung committed:
On installed app page, show apps if only team credential is installed

joeauyeung committed:
Clean up commented lines

joeauyeung committed:
Merge branch 'feat/organizations' into feat/apps-for-orgs

sean-brydon commented on apps/web/pages/api/integrations/[...args].ts:
```diff
@@ -29,21 +31,21 @@ const defaultIntegrationAddHandler = async ({
   req.session = await getServerSession({ req, res });
 
-  const { args } = req.query;
+  const { args, teamId } = req.query;
```
Shall we parse this with zod? Makes sure we dont have to cast as number below either (L:71)

sean-brydon commented on packages/app-store/_utils/createAppCredential.ts:
```diff
@@ -0,0 +1,57 @@
+  const state = decodeOAuthState(req);
+  if (state?.teamId) {
+    // Check that the user belongs to the team
+    const team = await prisma.team.findUnique({
+      where: { id: state.teamId },
+      select: { id: true, members: { select: { userId: true } } },
+    });
```
Havent tested but this should allow us to leave the search in the prisma query instead of in the IF below (L:31) 

```suggestion
    const team = await prisma.team.findUnique({
     where:{
      id:state.teamId,
      members:{
        some:{
          userId:ctx.user.id
        }
      }
    },
     select: { id: true, members: { select: { userId: true } } },
    });
```

joeauyeung replied:
Good catch, made the change.

joeauyeung replied:
@hariombalhara change made in `createOAuthAppCredentials`

sean-brydon commented on packages/app-store/_utils/createAppCredential.ts:
```diff
@@ -0,0 +1,57 @@
+      select: { id: true, members: { select: { userId: true } } },
+    });
+
+    if (team?.members.some((member) => member.userId === userId)) {
+      await prisma.credential.create({
+        data: {
+          type: appData.type,
+          key: key || {},
+          teamId: state.teamId,
+          appId: appData.appId,
+        },
+      });
+    } else {
+      throw new Error("User does not belong to the team");
+    }
```
If the above logic works we can early return 

if(!team) throw new Error etc

sean-brydon commented on packages/app-store/_utils/useAddAppMutation.ts:
```diff
@@ -57,9 +57,12 @@ function useAddAppMutation(_type: App["type"] | null, allOptions?: UseAddAppMuta
     };
     const stateStr = encodeURIComponent(JSON.stringify(state));
-    const searchParams = `?state=${stateStr}`;
+    const searchParams = `?state=${stateStr}${
+      variables && variables.teamId ? `&teamId=${variables.teamId}` : ""
+    }`;
```
More readable and cleaner imho 

```suggestion
const searchParams = new URLSearchParams({
  state: stateStr,
  ...(variables && variables.teamId && { teamId: variables.teamId }),
}).toString();
```

sean-brydon commented on packages/features/ee/teams/lib/getUserAdminTeams.ts:
```diff
@@ -0,0 +1,47 @@
+
+export default getUserAdminTeams;
+
+export type UserAdminTeams =
+  | (Awaited<Promise<ReturnType<typeof getUserAdminTeams>>>[number] & { isUser?: boolean })[]
+  | [];
```
Can be simplified a little bit. 

We dont need | [] from what i can see as even if there isnt any teams the typeof the array would still be 
`teams: {
    id: number;
    name: string;
    logo: string | null;
}[]
`
Just empty 

```suggestion
export type UserAdminTeams = (Awaited<ReturnType<typeof getUserAdminTeams>>[number] & { isUser?: boolean })[];

```

joeauyeung replied:
This wasn't picking up the property `isUser` so I decided to change this. Line 6 of the same file

sean-brydon reviewed (commented):
This is some amazing work joe. Left a few NITs. Impressed with how quick you got this together

joeauyeung committed:
Handle installing app to just an team event from event type page

leog committed:
Merge branch 'main' into feat/organizations

joeauyeung committed:
Fix early return when creating team app credential

joeauyeung committed:
Zoom add state to callback

joeauyeung committed:
Get team location credentials and save credential id to location

leog committed:
feat: Onboarding process to create an organization (#9184)

* Desktop first banner, mobile pending

* Removing dead code and img

* WIP

* Adds Email verification template+translations for organizations (#9202)

* First step done

* Merge branch 'feat/organizations-onboarding' of github.com:calcom/cal.com into feat/organizations-onboarding

* Step 2 done, avatar not working

* Covering null on unique clauses

* Onboarding admins step

* Last step to create teams

* Moving change password handler, improving verifying code flow

* Clearing error before submitting

* Reverting email testing api changes

* Reverting having the banner for now

* Consistent exported components

* Remove unneeded files from banner

* Removing uneeded code

* Fixing avatar selector

* Using meta component for head/descr

* Missing i18n strings

* Feedback

* Making an org avatar (temp)

* Check for subteams slug clashes with usernames

* Fixing create teams onsuccess

* feedback

* Making sure we check requestedSlug now

---------

Co-authored-by: sean-brydon <55134778+sean-brydon@users.noreply.github.com>

leog committed:
Merge branch 'main' into feat/organizations

leog committed:
Merge branch 'main' into feat/organizations

joeauyeung committed:
Type fix

joeauyeung committed:
Grab team location credentials

joeauyeung committed:
Add isInstalled to eventType apps query

leog committed:
feat: [CAL-1816] Organization subdomain support (#9345)

* Desktop first banner, mobile pending

* Removing dead code and img

* WIP

* Adds Email verification template+translations for organizations (#9202)

* First step done

* Merge branch 'feat/organizations-onboarding' of github.com:calcom/cal.com into feat/organizations-onboarding

* Step 2 done, avatar not working

* Covering null on unique clauses

* Onboarding admins step

* Last step to create teams

* Moving change password handler, improving verifying code flow

* Clearing error before submitting

* Reverting email testing api changes

* Reverting having the banner for now

* Consistent exported components

* Remove unneeded files from banner

* Removing uneeded code

* Fixing avatar selector

* Using meta component for head/descr

* Missing i18n strings

* Feedback

* Making an org avatar (temp)

* Check for subteams slug clashes with usernames

* Fixing create teams onsuccess

* Covering users and subteams, excluding non-org users

* Unpublished teams shows correctly

* Create subdomain in Vercel

* feedback

* Renaming Vercel env vars

* Vercel domain check before creation

* Supporting cal-staging.com

* Change to have vercel detect it

* vercel domain check data message error

* Remove check domain

* Making sure we check requestedSlug now

* Feedback and unneeded code

* Reverting unneeded changes

* Unneeded changes

---------

Co-authored-by: sean-brydon <55134778+sean-brydon@users.noreply.github.com>

leog committed:
Vercel subdomain creation in PROD only

joeauyeung committed:
Enable payment apps for team credentials

joeauyeung committed:
Merge branch 'feat/organizations' into feat/apps-for-orgs

alannnc committed:
Fix for team-user apps for event types

alannnc committed:
Merge with remote

leog committed:
Merge branch 'main' into feat/organizations

joeauyeung committed:
Fix layout and add teamId to app card

joeauyeung committed:
Disable apps on managed event types

joeauyeung committed:
Add managed event type fields to event type apps

leog committed:
Merge branch 'main' into feat/organizations

joeauyeung committed:
Include organizations in query

joeauyeung committed:
Merge branch 'feat/organizations' into feat/apps-for-orgs

joeauyeung committed:
Change createAppCredential to createOAuthAppCredential

joeauyeung commented on packages/features/bookings/lib/handleNewBooking.ts:
```diff
@@ -863,7 +863,10 @@ async function handler(
   const uid = translator.fromUUID(uuidv5(seed, uuidv5.URL));
 
-  const bookingLocation = getLocationValueForDB(locationBodyString, eventType.locations);
+  const { bookingLocation, bookingLocationCredentialId } = getLocationValueForDB(
```
If there is a `credentialId` then it will parse that from the event type location JSON

joeauyeung commented on packages/features/ee/managed-event-types/hooks/useLockedFieldsManager.tsx:
```diff
@@ -8,7 +8,7 @@ import type { _EventTypeModel } from "@calcom/prisma/zod/eventtype";
 import { Lock } from "@calcom/ui/components/icon";
 
-const Indicator = (label: string) => (
+export const LockedIndicator = (label: string) => (
```
Export this component and pass to event type app cards

alannnc commented on packages/app-store/giphy/api/add.ts:
```diff
@@ -10,15 +10,19 @@ import getInstalledAppPath from "../../_utils/getInstalledAppPath";
  * @param res
  */
 export default async function handler(req: NextApiRequest, res: NextApiResponse) {
+  console.log("🚀 ~ file: add.ts:13 ~ handler ~ req:", req);
```
```suggestion
```

alannnc commented on packages/app-store/giphy/lib/giphyManager.ts:
```diff
@@ -6,6 +6,7 @@ let api_key = "";
 
 const checkGiphyApiKey = async () => {
   const appKeys = await getAppKeysFromSlug("giphy");
+  console.log("🚀 ~ file: giphyManager.ts:9 ~ checkGiphyApiKey ~ appKeys:", appKeys);
```
```suggestion
```

alannnc commented on packages/app-store/routing-forms/api/add.ts:
```diff
@@ -9,12 +9,13 @@ const handler: AppDeclarativeHandler = {
         type: appType,
         key: {},
-        userId: user.id,
+        // userId: user.id,
```
```suggestion
       userId: user.id,
```

alannnc commented on packages/trpc/server/routers/loggedInViewer/deleteCredential.handler.ts:
```diff
@@ -27,12 +27,12 @@ type DeleteCredentialOptions = {
     where: {
       id: id,
-      userId: ctx.user.id,
+      OR: [{ userId: ctx.user.id }, { teamId }],
```
Don't use org here, if either it's provided just use that for the filter.
```suggestion
const whereDelete: Prisma.CredentialWhereUniqueInput  = {
   id,
};
if (teamId) {
whereDelete.teamId = teamId;
} else {
whereDelete.userId = ctx.user.id;
}

...
where: whereDelete,
...

```

alannnc commented on packages/trpc/server/routers/loggedInViewer/locationOptions.handler.ts:
```diff
@@ -4,22 +4,30 @@ import { getTranslation } from "@calcom/lib/server/i18n";
   const credentials = await prisma.credential.findMany({
     where: {
-      userId: ctx.user.id,
+      OR: [{ userId: ctx.user.id }, input?.teamId ? { teamId: input.teamId } : {}],
```
Here the same as above, adjust whereConditional if teamId is provided.

joeauyeung replied:
We're not including locations for V1 but will keep this in mind.

joeauyeung committed:
Show app installed on teams

leog committed:
Merge branch 'main' into feat/organizations

leog committed:
Making sure we let localhost still work

leog committed:
Merge branch 'main' into feat/organizations

joeauyeung committed:
UI show installed for which team

joeauyeung committed:
Type fixes

joeauyeung committed:
For team events move use host location to top

joeauyeung committed:
Add around to appStore

joeauyeung committed:
New team event types organizer default conf app

joeauyeung committed:
Fix app card bug

joeauyeung commented on packages/app-store/fathom/components/EventTypeAppCardInterface.tsx:
```diff
@@ -10,17 +9,17 @@ import type { appDataSchema } from "../zod";
+          updateEnabled(false);
         } else {
-          setEnabled(true);
+          updateEnabled(true);
```
Fixed in multiple app cards and in the app card template

joeauyeung committed:
Clean up

joeauyeung committed:
Merge branch 'feat/organizations' into feat/apps-for-orgs

joeauyeung committed:
Search for teamId or userId when deleting credential

joeauyeung committed:
Type fixes

joeauyeung committed:
Type fixes

joeauyeung committed:
Type fixes

joeauyeung committed:
Type fixes

sean-brydon commented on apps/web/components/AppListCard.tsx:
```diff
@@ -85,6 +88,21 @@ export default function AppListCard(props: AppListCardProps) {
+                <Avatar
+                  className="mr-2"
+                  alt={credentialOwner.name}
+                  size="sm"
```
NIT: I think 24px x 24px here might be a bit big 
```suggestion
                  size="xs"
```

sean-brydon commented on apps/web/components/eventtype/EventAppsTab.tsx:
```diff
@@ -62,11 +66,51 @@ export const EventAppsTab = ({ eventType }: { eventType: EventType }) => {
 
+  const appsWithTeamCredentials = eventTypeApps?.items.filter((app) => app.teams.length) || [];
+  const cardsForAppsWithTeams = appsWithTeamCredentials.map((app) => {
+    console.log("🚀 ~ file: EventAppsTab.tsx:104 ~ cardsForAppsWithTeams ~ app:", app);
```
```suggestion
```

sean-brydon commented on apps/web/pages/apps/installed/[category].tsx:
```diff
@@ -135,72 +137,106 @@ const IntegrationsList = ({ data, handleDisconnect, variant }: IntegrationsListP
+        invalidCredential={item?.invalidCredentialIds ? item.invalidCredentialIds.length > 0 : false}
+        credentialOwner={item?.credentialOwner}
+        actions={
+          <div className="flex  justify-end">
```
```suggestion
          <div className="flex justify-end">
```

sean-brydon commented on packages/app-store/_components/AppCard.tsx:
Nice changes :)

sean-brydon commented on packages/app-store/_utils/useIsAppEnabled.ts:
```diff
@@ -0,0 +1,56 @@
+
+import type { EventTypeAppCardApp } from "../types";
+
+const useIsAppEnabled = (app: EventTypeAppCardApp, getAppData: unknown, setAppData: unknown) => {
```
I managed to fix some of the type errors here (Might have knock on effects not sure) not blocking by any means 
```suggestion
import type { SetAppDataGeneric, GetAppDataGeneric } from "../EventTypeAppContext";
function useIsAppEnabled<TAppData extends ZodType>(
  app: EventTypeAppCardApp,
  getAppData: GetAppDataGeneric<TAppData>,
  setAppData: SetAppDataGeneric<TAppData>
){```

joeauyeung replied:
Where would `TAppData` come from? I'm thinking do we generate a schema file for each event type app (ex. `apps.keys-schemas.generated.ts` and pass it into `TAppData`?

sean-brydon replied:
As discussed in DMs TAppData is infered from usage on getAppData and setAppData

sean-brydon reviewed (commented):
Some minor nits 

Great work here @joeauyeung im amazed how quick you got this done! @hariombalhara It will 100% be worth you taking a look into this

joeauyeung committed:
Address feedback

leog committed:
Merge branch 'main' into feat/organizations

leog committed:
Feedback

leog committed:
Type check fixes

leog committed:
feat: Organization branding in side menu (#9279)

* Desktop first banner, mobile pending

* Removing dead code and img

* WIP

* Adds Email verification template+translations for organizations (#9202)

* First step done

* Merge branch 'feat/organizations-onboarding' of github.com:calcom/cal.com into feat/organizations-onboarding

* Step 2 done, avatar not working

* Covering null on unique clauses

* Onboarding admins step

* Last step to create teams

* Moving change password handler, improving verifying code flow

* Clearing error before submitting

* Reverting email testing api changes

* Reverting having the banner for now

* Consistent exported components

* Remove unneeded files from banner

* Removing uneeded code

* Fixing avatar selector

* Org branding provider used in shell sidebar

* Using meta component for head/descr

* Missing i18n strings

* Feedback

* Making an org avatar (temp)

* Using org avatar (temp)

* Not showing org logo if not set

* User onboarding with org branding (slug)

* Check for subteams slug clashes with usernames

* Fixing create teams onsuccess

* feedback

* Feedback

* Org public profile

* Public profiles for team event types

* Added setup profile alert

* Using org avatar on subteams avatar

* Making sure we show the set up profile on org only

* Profile username availability rely on org hook

* Update apps/web/pages/team/[slug].tsx

Co-authored-by: sean-brydon <55134778+sean-brydon@users.noreply.github.com>

* Update apps/web/pages/team/[slug].tsx

Co-authored-by: sean-brydon <55134778+sean-brydon@users.noreply.github.com>

---------

Co-authored-by: sean-brydon <55134778+sean-brydon@users.noreply.github.com>

leog committed:
Merge branch 'main' into feat/organizations

leog committed:
feat: Organization support for event types page (#9449)

* Desktop first banner, mobile pending

* Removing dead code and img

* WIP

* Adds Email verification template+translations for organizations (#9202)

* First step done

* Merge branch 'feat/organizations-onboarding' of github.com:calcom/cal.com into feat/organizations-onboarding

* Step 2 done, avatar not working

* Covering null on unique clauses

* Onboarding admins step

* Last step to create teams

* Moving change password handler, improving verifying code flow

* Clearing error before submitting

* Reverting email testing api changes

* Reverting having the banner for now

* Consistent exported components

* Remove unneeded files from banner

* Removing uneeded code

* Fixing avatar selector

* Org branding provider used in shell sidebar

* Using meta component for head/descr

* Missing i18n strings

* Feedback

* Making an org avatar (temp)

* Using org avatar (temp)

* Not showing org logo if not set

* User onboarding with org branding (slug)

* Check for subteams slug clashes with usernames

* Fixing create teams onsuccess

* feedback

* Feedback

* Org public profile

* Public profiles for team event types

* Added setup profile alert

* Using org avatar on subteams avatar

* Processing orgs and children as profile options

* Reverting change not belonging to this PR

* Making sure we show the set up profile on org only

* Removing console.log

* Comparing memberships to choose the highest one

---------

Co-authored-by: sean-brydon <55134778+sean-brydon@users.noreply.github.com>

leog committed:
Type errors

leog committed:
Merge branch 'main' into feat/organizations

zomars committed:
Refactor and type fixes

zomars committed:
Update orgDomains.ts

zomars committed:
Merge branch 'main' into feat/organizations

leog committed:
Feedback

leog committed:
Reverting

leog committed:
NIT

joeauyeung committed:
Merge branch 'feat/organizations' into feat/apps-for-orgs

joeauyeung committed:
Address feedback

leog committed:
Merge branch 'main' into feat/organizations

roae committed:
fix issue getting org slug from domain

leog committed:
Improving orgDomains util

leog committed:
Merge branch 'feat/organizations' of github.com:calcom/cal.com into feat/organizations

joeauyeung committed:
Merge branch 'feat/organizations' into feat/apps-for-orgs

leog committed:
Host comes with port

zomars committed:
Update useRouterQuery.ts

hariombalhara committed:
Merge remote-tracking branch 'origin/feat/organizations' into feat/apps-for-orgs

hariombalhara commented on packages/ui/components/apps/AppCard.tsx:
```diff
@@ -170,3 +167,103 @@ export function AppCard({ app, credentials, searchText }: AppCardProps) {
+    !userAdminTeams?.length ||
+    appCategories.some(
+      // We can remove these as we allow more of these installs
+      (category) => category === "calendar" || category === "video"
```
As per [this](https://github.com/calcom/cal.com/blob/d051a54aa91d4a34f96680405a18dfba39f04a82/apps/web/components/apps/App.tsx#L101) only `calendar` category allows multiple install.

joeauyeung replied:
Sorry this was an artifact comment. For orgs right now we don't want calendars and video apps to be installed team wide.

hariombalhara commented on packages/ui/components/apps/AppCard.tsx:
```diff
@@ -170,3 +167,103 @@ export function AppCard({ app, credentials, searchText }: AppCardProps) {
+    !userAdminTeams?.length ||
+    appCategories.some(
+      // We can remove these as we allow more of these installs
+      (category) => category === "calendar" || category === "video"
```
```suggestion
      (category) => category === "calendar"
```

hariombalhara commented on packages/app-store/_components/AppCard.tsx:
Yeah it should certainly be renamed. How about calling it BaseAppCard ? Can't think of a better name, atleast it would resolve the name conflict with ui/components/AppCard

hariombalhara commented on apps/web/components/AppListCard.tsx:
```diff
@@ -85,6 +88,21 @@ export default function AppListCard(props: AppListCardProps) {
+              <div className="flex items-center">
+                <Avatar
+                  className="mr-2"
+                  alt={credentialOwner.name || "Nameless"}
```
Followup: We have nameless spread across the app. I think I confirmed elsewhere that name is a required field and it shouldn't be unset. But even if that's the case, we should introduce `function getName({name:string})`

hariombalhara commented on apps/web/components/AppListCard.tsx:
```diff
@@ -19,6 +20,7 @@ type AppListCardProps = {
   isTemplate?: boolean;
   invalidCredential?: boolean;
   children?: ReactNode;
+  credentialOwner?: CredentialOwner;
```
I think we should make it a required field and expect `null` to be passed explicitly if someone doesn't want to pass `credentialOwner`. It helps in catching unexpected misses in prop passing.

I see CalendarListContainer(and other places) not passing the prop? Is that intentional?
<img width="747" alt="Screenshot 2023-06-14 at 6 05 21 PM" src="https://github.com/calcom/cal.com/assets/1780212/506cb1b7-dae6-4242-857e-33d240367bd1">

joeauyeung replied:
The places where we don't want to pass the `credentialOwner` prop are the apps where we don't want to be installed team-wide (calendars & videos for now).

joeauyeung committed:
Fix app card bug

zomars committed:
Merge remote-tracking branch 'origin/main' into feat/apps-for-orgs

PeerRich committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung committed:
Fix schema

joeauyeung committed:
Type fixes

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

zomars reviewed (dismissed):
e2e tests are failing due to new install app dropdown. Let's fix the code or update the test to take the dropdown into account.

joeauyeung committed:
Revert changes to location apps

joeauyeung committed:
Remove console.log

joeauyeung committed:
Fix app store test

joeauyeung commented:
> e2e tests are failing due to new install app dropdown. Let's fix the code or update the test to take the dropdown into account.

I fixed one test that was affected by this PR. The other tests are touching the bookers which are out of scope of this PR. By chance are other tests failing as well?

joeauyeung committed:
Handle install app dropdown

joeauyeung committed:
Add CalendarApp to `getCalendar`

joeauyeung commented on packages/app-store/_utils/getCalendar.ts:
```diff
@@ -19,8 +25,9 @@ export const getCalendar = async (credential: CredentialPayload | null): Promise
 
-  const calendarApp = await calendarAppImportFn();
-  if (!(calendarApp && "lib" in calendarApp && "CalendarService" in calendarApp.lib)) {
+  const calendarApp = (await calendarAppImportFn()) as CalendarApp;
+
+  if (!calendarApp.lib?.CalendarService) {
```
I added this new `CalendarApp` type as I got this weird type error.

https://www.loom.com/share/7dd62f03e6af463bb51605408b0a9e8c

zomars replied:
we can solve this with a type guard as well

zomars replied:
Type narrowing to be exact: https://www.typescriptlang.org/docs/handbook/2/narrowing.html

joeauyeung committed:
Add PaymentApp type fix

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung committed:
Payment type fix

joeauyeung committed:
Type fixes

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung commented on packages/trpc/server/routers/loggedInViewer/appCredentialsByType.handler.ts:
```diff
@@ -9,7 +11,34 @@ type AppCredentialsByTypeOptions = {
-  return user.credentials.filter((app) => app.type == input.appType).map((credential) => credential.id);
+  const userAdminTeams = await getUserAdminTeams({ userId: ctx.user.id, getUserInfo: true });
+
+  const teamIds = userAdminTeams.reduce((teamIds, team) => {
```
I'm getting a weird type error here. The type check thinks that `userAdminTeams` is type never but locally, it's being typed correctly.

![image](https://github.com/calcom/cal.com/assets/65426560/1f15e3c5-c6a2-487f-8ffb-20efdbf201a0)

joeauyeung committed:
Match with main

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung committed:
Change type to account for team

ericrommel reviewed (changes requested):
Labels from Team Event Type are not being updated after the app installations.

![image](https://github.com/calcom/cal.com/assets/10250819/e7930ea9-64f9-405c-ab9c-efe8a8a85228)

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung committed:
Fix app count for team events

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung commented:
> Labels from Team Event Type are not being updated after the app installations.
> 
> ![image](https://user-images.githubusercontent.com/10250819/248053122-e7930ea9-64f9-405c-ab9c-efe8a8a85228.png)

@ericrommel this is now fixed

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung committed:
Type fixes

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung committed:
More type fixes

joeauyeung committed:
Type fix?

joeauyeung committed:
Fix the type fix

joeauyeung committed:
Remove UserAdminTeams empty array union

keithwillcode committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung committed:
Type fix

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung committed:
Type fix

joeauyeung committed:
Type fix

ericrommel reviewed (changes requested):
The form for new orgs shows `undefined` as the default value for `Organization URL`. Also, after filling out the `Your organization email address` field, the `Organization URL` doesn't change.

![image](https://github.com/calcom/cal.com/assets/10250819/f9defb15-8b11-4c4e-ab44-06fc603f6eeb)

ericrommel reviewed (dismissed):
Apps already installed are not being shown on the Event Type area, under app installed. E.g., Fathon is already installed, but when adding an Event Type, it's not. However, if you try to install it again, you get a notification saying that it's already installed.
![image](https://github.com/calcom/cal.com/assets/10250819/c2def92c-4927-4b7b-8615-2ce72bb2e72f)

![image](https://github.com/calcom/cal.com/assets/10250819/1e326da3-49a5-4360-988e-8eddf0641132)

zomars commented on apps/web/components/eventtype/EventAppsTab.tsx:
```diff
@@ -92,15 +125,18 @@ export const EventAppsTab = ({ eventType }: { eventType: EventType }) => {
+          {cardsForAppsWithTeams.map((apps) => apps.map((cards) => cards))}
+          {installedApps
+            ?.filter((app) => !app.teams.length)
+            .map((app) => (
```
@sean-brydon you're kinda right. But this code would leave us with a mixed array of teams and undefined values.

zomars commented on apps/web/components/eventtype/EventAppsTab.tsx:
```diff
@@ -16,12 +16,16 @@ export type EventType = Pick<EventTypeSetupProps, "eventType">["eventType"] &
   const methods = useFormContext<FormValues>();
-  const installedApps = eventTypeApps?.filter((app) => app.credentials.length);
-  const notInstalledApps = eventTypeApps?.filter((app) => !app.credentials.length);
+  const installedApps =
+    eventTypeApps?.items.filter((app) => app.credentialIds.length || app.teams.length) || [];
+  const notInstalledApps =
+    eventTypeApps?.items.filter((app) => !app.credentialIds.length && !app.teams.length) || [];
```
Got for it!

zomars committed:
Merge branch 'main' into feat/apps-for-orgs

zomars committed:
Uses type predicates

zomars commented on apps/web/components/apps/App.tsx:
```diff
@@ -385,3 +386,97 @@ export default function App(props: {
+          {userAdminTeams.map((team) => (
+            <DropdownItem
+              type="button"
+              data-testid={team.isUser ? "install-app-button-personal" : "anything else"}
```
Why do we need to do this? I feel that we're introducing unnecessary bloat here. It's hard to understand why do we need this by reading the code itself.

joeauyeung replied:
It's a part of the first iteration for orgs until the account switcher is created. @Jaibles

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

hariombalhara commented on packages/lib/constants.ts:
```diff
@@ -8,7 +8,7 @@ export const WEBAPP_URL =
   HEROKU_URL ||
   RENDER_URL ||
-  "http://localhost:3000";
+  "http:/localhost:3000";
```
```suggestion
  "http://localhost:3000";
```

joeauyeung replied:
Added after merge with `main`

hariombalhara commented on packages/ui/components/apps/AppCard.tsx:
```diff
@@ -170,3 +167,100 @@ export function AppCard({ app, credentials, searchText }: AppCardProps) {
+              }}>
+              <p>
+                {team.name}{" "}
+                {credentials &&
+                  credentials.some((credential) =>
+                    credential?.teamId ? credential?.teamId === team.id : credential.userId === team.id
```
It would be better to extract it out in a variable called `isInstalledForTeamOrUser` as it's used in 2 places.

hariombalhara commented:
Not able to install an app to multiple teams. After installing it for user, I can't install it for a team.
https://www.loom.com/share/8e86d8fc5dc04b8f9b86c658adbaf441

hariombalhara committed:
Use teamId. Fixes installation for teams after user installation

hariombalhara commented:
> Not able to install an app to multiple teams. After installing it for user, I can't install it for a team. [loom.com/share/8e86d8fc5dc04b8f9b86c658adbaf441](https://www.loom.com/share/8e86d8fc5dc04b8f9b86c658adbaf441)

Fixed it. It was a simple non-use of teamId which I think started happening after some change. cc @joeauyeung I would be reviewing, testing and fixing things that I can.

hariombalhara commented:
On EventType - Apps Tab, an app for a Team Event(created for an organization's(VWO) Team(Test Team)) is shown twice.
![image](https://github.com/calcom/cal.com/assets/1780212/a14622e9-9aae-4bb2-b39b-d1125e886166)

Fathom app is installed for both VWO(organization) and Test Team(a team of VWO organization) and it is shown twice due to it.
<img width="1010" alt="Screenshot 2023-06-29 at 4 40 27 PM" src="https://github.com/calcom/cal.com/assets/1780212/02dde194-599b-48cb-8cee-a84b59f09034">

hariombalhara committed:
Fix Team Events not working

hariombalhara committed:
Get embed for org events working

hariombalhara committed:
Fix rewrites

hariombalhara committed:
Merge remote-tracking branch 'origin/org-team-event-fix' into feat/apps-for-orgs

hariombalhara commented on packages/app-store/_utils/createOAuthAppCredential.ts:
```diff
@@ -0,0 +1,64 @@
+        id: state.teamId,
+        members: {
+          some: {
+            userId: req.session?.user.id,
```
```suggestion
            userId: req.session?.user.id,
            accepted: true
```
We are not verifying the accepted status. So, members in pending invitation state might be able to create credentials for team.

hariombalhara commented on packages/app-store/_utils/useAddAppMutation.ts:
```diff
@@ -57,9 +57,13 @@ function useAddAppMutation(_type: App["type"] | null, allOptions?: UseAddAppMuta
             location.search
           ),
       ...(type === "google_calendar" && { installGoogleVideo: options?.installGoogleVideo }),
+      ...(variables && variables.teamId && { teamId: variables.teamId }),
```
nit: Better to save variables.teamId in a variable and reuse that.

hariombalhara commented on packages/app-store/_utils/useIsAppEnabled.ts:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    const credentialId = getAppData("credentialId");
+
+    // For pre-org team event types assume that the app used was the user's if enabled
+    if (!credentialId) {
+      if (!app.credentialOwner?.teamId) {
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        setEnabled(getAppData("enabled"));
+      }
+    }
+    // App type is not being picked up in the type definition
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    if (app?.credentialIds && app.credentialIds.some((id) => id === credentialId)) {
+      setEnabled(true);
+    }
+    return;
+  }, []);
+
+  // If the app is enabled then we should also write the credentialId to the app metadata
+  useEffect(() => {
+    if (enabled && app.credentialIds) {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
+      setAppData("credentialId", app.credentialIds[0]);
+    }
+    if (!enabled) {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
+      setAppData("credentialId", undefined);
+    }
+  }, [enabled]);
+
+  const updateEnabled = (newValue: boolean) => {
+    setEnabled(newValue);
+  };
```
I feel this logic can be heavily simplified like this. It also avoids useEffect which are always the last resort.
```Javascript
const [enabled, setEnabled] = useState(() => {
    if (!app.credentialOwner) {
      return getAppData("enabled");
    }
    const credentialId = getAppData("credentialId");
    const isAppEnabledForCredential =
      app?.credentialIds && app.credentialIds.some((id) => id === credentialId);
    return isAppEnabledForCredential;
  });

  const updateEnabled = (newValue: boolean) => {
    if (!newValue) {
      setAppData("credentialId", undefined);
    }

    if (newValue && app.credentialIds) {
      setAppData("credentialId", app.credentialIds[0]);
    }
    setEnabled(newValue);
  };
```

hariombalhara commented on packages/app-store/_utils/useIsAppEnabled.ts:
```diff
@@ -0,0 +1,62 @@
+
+function useIsAppEnabled<TAppData extends ZodType>(
+  app: EventTypeAppCardApp,
+  getAppData: GetAppDataGeneric<TAppData>,
+  setAppData: SetAppDataGeneric<TAppData>
```
I think instead of passing these as args we should be able to get them from `useAppContextWithSchema`

hariombalhara commented on packages/app-store/ga4/components/EventTypeAppCardInterface.tsx:
```diff
@@ -1,32 +1,36 @@
+const EventTypeAppCard: EventTypeAppCardComponent = function EventTypeAppCard({
+  app,
+  eventType,
+  disabled,
+  LockedIcon,
```
Can we get these from useAppContextWithSchema itself ?

joeauyeung replied:
Great idea, moved these into the provider in this commit b2826f5497df8b671f405c8bd0f768b0d8c9a88e

hariombalhara commented on packages/app-store/utils.ts:
```diff
@@ -57,10 +57,15 @@ export function getLocationGroupedOptions(integrations: ReturnType<typeof getApp
-        app.categories.length >= 2 ? app.categories.find((category) => category !== "video") : app.category;
-      if (!category) category = "video";
-      const option = { ...app.locationOption, icon: app.logo, slug: app.slug };
+      const nonVideoCategories = app.categories.find((category) => category !== "video");
```
```suggestion
      const hasNonVideoCategory = app.categories.find((category) => category !== "video");
```

joeauyeung replied:
Merge conflict with `main` didn't have these lines

hariombalhara commented on packages/app-store/zoomvideo/api/add.ts:
```diff
@@ -19,11 +20,13 @@ async function handler(req: NextApiRequest) {
   });
 
   const { client_id } = await getZoomAppKeys();
+  const state = encodeOAuthState(req);
```
nit: I find it a more apt name `getOAuthStateAsJson`

hariombalhara commented on packages/core/EventManager.ts:
```diff
@@ -371,7 +371,7 @@ export default class EventManager {
    */
 
-  private getVideoCredential(event: CalendarEvent): CredentialWithAppName | undefined {
+  private async getVideoCredential(event: CalendarEvent): Promise<CredentialWithAppName | undefined> {
```
Why is it declared async now without any change in functionality?

joeauyeung replied:
I did have locations enabled for this PR, but we realized it will take more design work. There are some remnants of that functionality that I may have missed.

hariombalhara commented on packages/features/auth/lib/getServerSession.ts:
```diff
@@ -46,7 +46,7 @@ export async function getServerSession(options: {
   }
 
-  const user = await prisma.user.findUnique({
+  const user = await prisma.user.findFirst({
```
Why are we doing a findFirst. emails is still unique field. Looks like a change from one of the org branches that was reverted before going to main.

joeauyeung replied:
Merged with `main`

hariombalhara commented on packages/features/auth/lib/next-auth-options.ts:
```diff
@@ -69,7 +69,7 @@ const providers: Provider[] = [
       }
 
-      const user = await prisma.user.findUnique({
+      const user = await prisma.user.findFirst({
```
This too should be reverted.

hariombalhara commented on packages/features/bookings/lib/handleCancelBooking.ts:
```diff
@@ -593,8 +596,10 @@ async function handler(req: CustomRequest) {
     // Posible to refactor TODO:
-    const paymentApp = await appStore[paymentAppCredential?.app?.dirName as keyof typeof appStore]();
-    if (!(paymentApp && "lib" in paymentApp && "PaymentService" in paymentApp.lib)) {
+    const paymentApp = (await appStore[
+      paymentAppCredential?.app?.dirName as keyof typeof appStore
+    ]()) as PaymentApp;
+    if (!paymentApp.lib?.PaymentService) {
```
Why are we doing a type assertion? Earlier code was safer as it's possible that paymentApp or paymentApp.lib is nullish

joeauyeung replied:
I was getting a weird type error in this file. Here's a Loom with what I was experiencing. Still the same app.

https://www.loom.com/share/7dd62f03e6af463bb51605408b0a9e8c

hariombalhara commented on packages/features/bookings/lib/handleNewBooking.ts:
```diff
@@ -1559,7 +1559,9 @@ async function handler(
         const credentialPaymentAppCategories = await prisma.credential.findMany({
           where: {
-            userId: organizerUser.id,
+            ...(paymentAppData.credentialId
+              ? { id: paymentAppData.credentialId }
+              : { userId: organizerUser.id }),
```
nit: I think we could just use `credentialId` in both team App and non team app installation case. It would simplify the code.

joeauyeung replied:
This was for backwards compatibility if current `paymentAppData` doesn't contain the `credentialId`

hariombalhara replied:
Ah makes sense.

hariombalhara commented on apps/web/components/apps/App.tsx:
```diff
@@ -385,3 +386,97 @@ export default function App(props: {
+                );
+              }}>
+              <p>{team.name}{" "}
+                {credentials &&
+                  credentials.some((credential) =>
+                    credential?.teamId ? credential?.teamId === team.id : credential.userId === team.id
+                  ) &&
```
Should be abstracted it into a variable like `isInstalled`.

hariombalhara commented on apps/web/components/ui/form/LocationSelect.tsx:
```diff
@@ -40,11 +42,13 @@ export default function LocationSelect(props: Props<LocationOption, false, Group
+              <OptionWithIcon icon={props.data.icon} label={props.data.label} />
+            </components.Option>
+          );
+        },
```
Looks like we are not supporting credentialId in here. We can revert this change.

Also @joeauyeung are we planning to enable Video Apps for teams with this PR?

joeauyeung replied:
Adding video apps will be in a separate PR as it requires more designs.

hariombalhara commented on packages/trpc/server/routers/loggedInViewer/integrations.handler.ts:
```diff
@@ -12,21 +14,57 @@ type IntegrationsOptions = {
   let apps = enabledApps.map(
     ({ credentials: _, credential: _1, key: _2 /* don't leak to frontend */, ...app }) => {
-      const credentialIds = credentials.filter((c) => c.type === app.type).map((c) => c.id);
+      const credentialIds = credentials.filter((c) => c.type === app.type && !c.teamId).map((c) => c.id);
```
Either we should not filter out team credentials or we should rename it to `userCredentialIds`. It is confusing as is.

joeauyeung replied:
@hariombalhara I changed the the variable to `userCredentialIds`. I searched where `integrations.useQuery()` was used to make sure the change didn't affect any functionality. You can see the commit here 08ff16d500f2c54340ac8686a6a30c1b942d5829

hariombalhara commented on packages/trpc/server/routers/loggedInViewer/deleteMeWithoutPassword.handler.ts:
```diff
@@ -12,7 +12,7 @@ type DeleteMeWithoutPasswordOptions = {
 
 export const deleteMeWithoutPasswordHandler = async ({ ctx }: DeleteMeWithoutPasswordOptions) => {
-  const user = await prisma.user.findUnique({
+  const user = await prisma.user.findFirst({
```
```suggestion
  const user = await prisma.user.findUnique({
```

hariombalhara replied:
@joeauyeung This and some other comments are marked resolved but there aren't any changes. Can you add a comment if it's not supposed to change before resolving the comment.
If you have made the suggested change you can just resolve the conversation.

joeauyeung replied:
With the comments without any changes, it looks like it synced properly with `main` so there wouldn't be any changes reflected in this PR.

hariombalhara commented on packages/trpc/server/routers/loggedInViewer/integrations.handler.ts:
```diff
@@ -44,8 +124,20 @@ export const integrationsHandler = async ({ ctx, input }: IntegrationsOptions) =
+  }
+
+  if (extendsFeature) {
+    apps = apps
+      .filter((app) => app.extendsFeature?.includes(extendsFeature))
+      .map((app) => ({
+        ...app,
+        isInstalled: !!app.credentialIds?.length || !!app.teams?.length || app.isGlobal,
+      }));
```
What's the significance of this code? I don't think it's being used.

hariombalhara commented on apps/web/components/eventtype/EventAppsTab.tsx:
```diff
@@ -16,12 +16,16 @@ export type EventType = Pick<EventTypeSetupProps, "eventType">["eventType"] &
 
 export const EventAppsTab = ({ eventType }: { eventType: EventType }) => {
   const { t } = useLocale();
-  const { data: eventTypeApps, isLoading } = trpc.viewer.apps.useQuery({
```
This makes viewer.apps a dead code. We can remove that.

hariombalhara commented on packages/trpc/server/routers/loggedInViewer/integrations.handler.ts:
```diff
@@ -10,23 +13,100 @@ type IntegrationsOptions = {
+
+  if (includeTeamInstalledApps || teamId) {
+    // Get app credentials that the user is an admin or owner to
+    userAdminTeams = await prisma.team.findMany({
```
nit: Good candidate for getUserAdminTeams

joeauyeung replied:
Good point. This query included the parent team so I add a `getParentInfo` to `getUserAdminTeams`

hariombalhara commented on packages/trpc/server/routers/loggedInViewer/integrations.handler.ts:
```diff
@@ -10,23 +13,100 @@ type IntegrationsOptions = {
+        ...(teamId && { id: teamId }),
+        members: {
+          some: {
+            userId: user.id,
```
```suggestion
            userId: user.id,
            accepted: true
```

hariombalhara commented:
Non admin/owner team member not able to see the App installed for it's Team or Org.
- Installed Apps View
<img width="1419" alt="Screenshot 2023-07-03 at 10 16 39 AM" src="https://github.com/calcom/cal.com/assets/1780212/da9ba7fe-4882-4a30-a455-a72114424545">
<img width="1559" alt="Screenshot 2023-07-03 at 10 16 50 AM" src="https://github.com/calcom/cal.com/assets/1780212/3521b2a3-7432-4a12-a95d-155a3f50a532">

hariombalhara reviewed (changes requested):
So many complexities and changes in this PR. Great work so far @joeauyeung 🙏.
I have left some comments. The feedback that's important to be addressed, I have added in the PR itself so that it doesn't get lost with so many comments.

joeauyeung commented:
> On EventType - Apps Tab, an app for a Team Event(created for an organization's(VWO) Team(Test Team)) is shown twice. ![image](https://user-images.githubusercontent.com/1780212/249753938-a14622e9-9aae-4bb2-b39b-d1125e886166.png)
> 
> Fathom app is installed for both VWO(organization) and Test Team(a team of VWO organization) and it is shown twice due to it. <img alt="Screenshot 2023-06-29 at 4 40 27 PM" width="1010" src="https://user-images.githubusercontent.com/1780212/249754494-02dde194-599b-48cb-8cee-a84b59f09034.png">



> On EventType - Apps Tab, an app for a Team Event(created for an organization's(VWO) Team(Test Team)) is shown twice. ![image](https://user-images.githubusercontent.com/1780212/249753938-a14622e9-9aae-4bb2-b39b-d1125e886166.png)
> 
> Fathom app is installed for both VWO(organization) and Test Team(a team of VWO organization) and it is shown twice due to it. <img alt="Screenshot 2023-06-29 at 4 40 27 PM" width="1010" src="https://user-images.githubusercontent.com/1780212/249754494-02dde194-599b-48cb-8cee-a84b59f09034.png">

This is done on purpose. There could be an instance where there is the same app but an organization scoped credential maybe different from a team scoped credential.

hariombalhara commented on packages/app-store/gtm/zod.ts:
```diff
@@ -4,12 +4,15 @@ import { eventTypeAppCardZod } from "@calcom/app-store/eventTypeAppCardZod";
+        trackingId = !trackingId.startsWith("GTM-") ? `GTM-${trackingId}` : trackingId;
+        return trackingId;
+      })
+      .optional(),
```
What's the need to make it optional?

joeauyeung committed:
Address feedback

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung committed:
Type fix

joeauyeung committed:
Fixes

joeauyeung committed:
Add useAppContextWithSchema in useIsAppEnabled

joeauyeung committed:
Type fix for apps using useIsAppEnabled

joeauyeung committed:
Integrations.handler change credentialIds to userCredentialIds

joeauyeung committed:
Remove apps endpoint

joeauyeung committed:
Add LockedIcon and disabled props to event type app context

joeauyeung committed:
Type fixes

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung committed:
Type fix

joeauyeung committed:
Type fixes

joeauyeung committed:
Show team installed apps for members

joeauyeung commented:
@hariombalhara installed apps for teams will now appear for members. These apps will be readonly. Commit 
40e20dcaa196ba94dfb0e5a18ac775af5a73535d

![image](https://github.com/calcom/cal.com/assets/65426560/476a0bc7-dee3-4294-8596-2ea112df3f56)

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung committed:
Type fix

hariombalhara commented on packages/features/ee/sso/lib/saml.ts:
```diff
@@ -31,7 +31,7 @@ export const isSAMLAdmin = (email: string) => {
 
 export const samlTenantProduct = async (prisma: PrismaClient, email: string) => {
-  const user = await prisma.user.findUnique({
+  const user = await prisma.user.findFirst({
```
```suggestion
  const user = await prisma.user.findUnique({
```

hariombalhara committed:
Reverting findFirst

hariombalhara commented on packages/trpc/server/routers/loggedInViewer/deleteMe.handler.ts:
```diff
@@ -20,7 +20,7 @@ type DeleteMeOptions = {
 export const deleteMeHandler = async ({ ctx, input }: DeleteMeOptions) => {
   // Check if input.password is correct
-  const user = await prisma.user.findUnique({
+  const user = await prisma.user.findFirst({
```
```suggestion
  const user = await prisma.user.findUnique({
```

hariombalhara committed:
Revert findFirst

hariombalhara committed:
Merge remote-tracking branch 'origin/feat/apps-for-orgs' into feat/apps-for-orgs

hariombalhara commented on packages/features/bookings/lib/handleCancelBooking.ts:
```diff
@@ -593,8 +596,10 @@ async function handler(req: CustomRequest) {
+    const paymentApp = (await appStore[
+      paymentAppCredential?.app?.dirName as keyof typeof appStore
+    ]()) as PaymentApp;
+    if (!paymentApp.lib?.PaymentService) {
```
```suggestion
    if (!paymentApp?.lib?.PaymentService) {
```
That type error is seriously weird. Not focussing on it, just making the code runtime error free

hariombalhara committed:
Avoid a possible 500

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into feat/apps-for-orgs

hariombalhara committed:
Fix missing tanslation

hariombalhara reviewed (commented):
QR Code App went crazy. I have it installed just once for Organization. Let's fix it in a followup.

![Screenshot 2023-07-05 at 1.26.29 PM.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/QOdRMCCW7wQwTOiuNnjo/34e3d1f6-193f-4432-ae57-44322d285466/Screenshot%202023-07-05%20at%201.26.29%20PM.png)

**Org Owner View**
![Screenshot 2023-07-05 at 1.27.00 PM.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/QOdRMCCW7wQwTOiuNnjo/76b948a9-72a2-40d3-932c-72af364ec9fd/Screenshot%202023-07-05%20at%201.27.00%20PM.png)

hariombalhara committed:
Merge remote-tracking branch 'origin/feat/apps-for-orgs' into feat/apps-for-orgs

hariombalhara committed:
Avoid possible 500

hariombalhara reviewed (dismissed):
The biggest PR I have ever reviewed 🙏 Thanks @joeauyeung for the quick implementation of feedback. 
I have tested org stuff and non-org stuff and feel confident enough to merge it. There might still be bugs but we can do followup for that.

joeauyeung committed:
Undo default app for teams

joeauyeung commented on packages/trpc/server/routers/viewer/eventTypes/create.handler.ts:
```diff
@@ -40,8 +40,11 @@ export const createHandler = async ({ ctx, input }: CreateOptions) => {
-  if (defaultConferencingData && defaultConferencingData.appSlug !== "daily-video") {
+  // If it is a team event default to the organizer's default conferencing app
+  if (teamId) {
+    locations = [{ type: "conferencing" }];
```
It was going to small fix I was putting into the PR but it's going to need more work for managed event types. I'm removing the change and creating a separate PR for this.

joeauyeung commented on packages/app-store/fathom/components/EventTypeAppCardInterface.tsx:
```diff
@@ -1,32 +1,31 @@
     <AppCard
       setAppData={setAppData}
       app={app}
+      disableSwitch={disabled}
```
I found some other improvements we can make, so I'm creating a separate PR.

https://github.com/calcom/cal.com/pull/9950

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung committed:
Type fix

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs

hariombalhara committed:
Fix test

PeerRich committed:
Update package.json

sean-brydon committed:
feat: Fix invite bug - added tests (#9945)

Co-authored-by: Hariom Balhara <hariombalhara@gmail.com>

sean-brydon committed:
chore: Button Component Tidy up (#9888)

Co-authored-by: Peer Richelsen <peeroke@gmail.com>

Udit-takkar committed:
feat: Make Team Private

## What does this PR do?

Fixes https://github.com/calcom/cal.com/issues/8974

1) When user is admin

<img width="1440" alt="Screenshot 2023-07-03 at 6 45 50 PM" src="https://github.com/calcom/cal.com/assets/53316345/ce15158f-d278-4f1a-ba2e-8b63e4274793">

2) When user is not admin and team is private

<img width="1440" alt="Screenshot 2023-07-03 at 6 47 15 PM" src="https://github.com/calcom/cal.com/assets/53316345/ce23560e-690a-4c42-a76d-49691260aa4d">

3) 
<img width="1440" alt="Screenshot 2023-07-03 at 6 51 56 PM" src="https://github.com/calcom/cal.com/assets/53316345/13af38f8-5618-4dae-b359-b24dc91e4eb4">


## Type of change

<!-- Please delete bullets that are not relevant. -->

- New feature (non-breaking change which adds functionality)

## How should this be tested?

1) go to Team members page and turn on switch Make Team Private.

Now after making the team private only admin would be able to see all the members list in the settings. There will not be a button to Book a team member instead on the team page like before.


## Mandatory Tasks

- [ ] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

joeauyeung committed:
Merge remote-tracking branch 'origin/main' into feat/apps-for-orgs

joeauyeung committed:
Merge branch 'main' into feat/apps-for-orgs
